rule-engine/rule-engine-components/src/test/java/org/thingsboard/rule/engine/geo/TbGpsGeofencingActionNodeTest.java

```diff
@@ -0,0 +1,269 @@
+                // default config with presenceMonitoringStrategyOnEachMessage false and msgInside false
+                Arguments.of(new GpsGeofencingActionTestCase(deviceId, false, false, new EntityGeofencingState(false, 0, false)), LEFT),
+                Arguments.of(new GpsGeofencingActionTestCase(deviceId, false, false, new EntityGeofencingState(false, tsNow, false)), SUCCESS),
+                Arguments.of(new GpsGeofencingActionTestCase(deviceId, false, false, new EntityGeofencingState(false, tsNowMinusMinuteAndMillis, false)), OUTSIDE),
+                Arguments.of(new GpsGeofencingActionTestCase(deviceId, false, false, new EntityGeofencingState(false, tsNow, true)), SUCCESS),
+                // default config with presenceMonitoringStrategyOnEachMessage true and msgInside true
+                Arguments.of(new GpsGeofencingActionTestCase(deviceId, true, true, new EntityGeofencingState(false, 0, false)), ENTERED),
+                Arguments.of(new GpsGeofencingActionTestCase(deviceId, true, true, new EntityGeofencingState(true, tsNow, false)), INSIDE),
+                Arguments.of(new GpsGeofencingActionTestCase(deviceId, true, true, new EntityGeofencingState(true, tsNowMinusMinuteAndMillis, false)), INSIDE),
+                // default config with presenceMonitoringStrategyOnEachMessage true and msgInside false
+                Arguments.of(new GpsGeofencingActionTestCase(deviceId, false, true, new EntityGeofencingState(false, 0, false)), LEFT),
+                Arguments.of(new GpsGeofencingActionTestCase(deviceId, false, true, new EntityGeofencingState(false, tsNow, false)), OUTSIDE),
+                Arguments.of(new GpsGeofencingActionTestCase(deviceId, false, true, new EntityGeofencingState(false, tsNowMinusMinuteAndMillis, false)), OUTSIDE)
+        );
+    }
+
+    @ParameterizedTest
+    @MethodSource
+    void givenReportPresenceStatusOnEachMessage_whenOnMsg_thenVerifyOutputMsgType(
+            GpsGeofencingActionTestCase gpsGeofencingActionTestCase,
+            String expectedOutput
+    ) throws TbNodeException {
+        // GIVEN
+        var config = new TbGpsGeofencingActionNodeConfiguration().defaultConfiguration();
+        config.setReportPresenceStatusOnEachMessage(gpsGeofencingActionTestCase.isReportPresenceStatusOnEachMessage());
+
+        node.init(ctx, new TbNodeConfiguration(JacksonUtil.valueToTree(config)));
+
+        TbMsg msg = gpsGeofencingActionTestCase.isMsgInside() ?
+                getInsideRectangleTbMsg(gpsGeofencingActionTestCase.getEntityId()) :
+                getOutsideRectangleTbMsg(gpsGeofencingActionTestCase.getEntityId());
+
+        when(ctx.getAttributesService()).thenReturn(attributesService);
+        when(ctx
+                .getAttributesService()
+                .find(ctx.getTenantId(), msg.getOriginator(), DataConstants.SERVER_SCOPE, ctx.getServiceId()))
+                .thenReturn(Futures.immediateFuture(Optional.empty()));
```

Comment:
```suggestion
        when(ctx.getAttributesService()
                .find(ctx.getTenantId(), msg.getOriginator(), DataConstants.SERVER_SCOPE, ctx.getServiceId()))
                .thenReturn(Futures.immediateFuture(Optional.empty()));
```

Reply:
I think that it's better to say here any() for each argument of the find method call. And then verify that it's called once with correct values after onMsg is called.

```
// create tenantId and serviceId final fields in top of test class and then:
when(ctx.getTenantId()).thenReturn(tenantId);
when(ctx.getServiceId()).thenReturn(serviceId);
when(ctx.getAttributesService()
                .find(any(), any(), any(), any()))
                .thenReturn(Futures.immediateFuture(Optional.empty()));
```                

and then

```
verify(ctx).getAttributesService()
                .find(eq(tenantId), eq(msg.getOriginator()), eq(DataConstants.SERVER_SCOPE), eq(serviceId)));
```